File: src/2370. Longest Ideal Subsequence.cpp
class Solution {
    public:
        int longestIdealString(string s, int k) {
            int dp[26] = {};
            int maxCount = 0;
            for (char & c: s) {
                int i = c - 'a';
                int minIndex = max(i - k, 0);
                int maxIndex = min(i + k, 25);
                for (int j = minIndex; j <= maxIndex; ++j) {
                    dp[i] = max(dp[i], dp[j]);
                }
                maxCount = max(maxCount, ++dp[i]);
            }
            return maxCount;
        }
};

File: src/2370. Longest Ideal Subsequence.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
export function longestIdealString(s, k) {
  /**
   * dp[i] means max subsequence count for specific alphabet end.
   * 0 -> a
   * ...
   * 25 -> z
   */
  const dp = Array(26).fill(0)
  for (let char of s) {
    const i = char.charCodeAt(0) - 97
    let minIndex = Math.max(i - k, 0),
      maxIndex = Math.min(i + k, 25)
    for (let j = minIndex; j <= maxIndex; j++) {
      dp[i] = Math.max(dp[j], dp[i])
    }
    dp[i]++
  }
  return dp.reduce((maxCount, count) => Math.max(maxCount, count), 0)
}
